langchain/document_loaders/generic.py

```diff
@@ -0,0 +1,131 @@
+
+from langchain.document_loaders.base import BaseBlobParser, BaseLoader
+from langchain.document_loaders.blob_loaders import BlobLoader, FileSystemBlobLoader
+from langchain.document_loaders.parsers.registry import get_parser
+from langchain.schema import Document
+from langchain.text_splitter import TextSplitter
+
+_PathLike = Union[str, Path]
+
+
+class GenericLoader(BaseLoader):
+    """A generic document loader.
+
+    A generic document loader that allows combining an arbitrary blob loader with
+    a blob parser.
+
+    Examples:
+
+        .. code-block:: python
+
+        from langchain.document_loaders import GenericLoader
+        from langchain.document_loaders.blob_loaders import FileSystemBlobLoader
+
+        loader = GenericLoader.from_filesystem(
+            path="path/to/directory",
+            glob="**/[!.]*",
+            suffixes=[".pdf"],
+            show_progress=True,
+        )
+
+        docs = loader.lazy_load()
+        next(docs)
+
+        Example instantiations to change which files are loaded:
+
+        ... code-block:: python
+
+            # Recursively load all text files in a directory.
+            loader = GenericLoader.from_filesystem("/path/to/dir", glob="**/*.txt")
+
+            # Recursively load all non-hidden files in a directory.
+            loader = GenericLoader.from_filesystem("/path/to/dir", glob="**/[!.]*")
+
+            # Load all files in a directory without recursion.
+            loader = GenericLoader.from_filesystem("/path/to/dir", glob="*")
+
+        Example instantiations to change which parser is used:
+
+        ... code-block:: python
+
+            from langchain.document_loaders.parsers.pdf import PyPDFParser
+
+            # Recursively load all text files in a directory.
+            loader = GenericLoader.from_filesystem(
+                "/path/to/dir",
+                glob="**/*.pdf",
+                parser=PyPDFParser()
+            )
+    """
+
+    def __init__(
+        self,
+        blob_loader: BlobLoader,
+        blob_parser: BaseBlobParser,
+    ) -> None:
+        """A generic document loader.
+
+        Args:
+            blob_loader: A blob loader which knows how to yield blobs
+            blob_parser: A blob parser which knows how to parse blobs into documents
+        """
+        self.blob_loader = blob_loader
+        self.blob_parser = blob_parser
+
+    def lazy_load(
+        self,
+    ) -> Iterator[Document]:
+        """Load documents lazily. Use this when working at a large scale."""
+        for blob in self.blob_loader.yield_blobs():
+            yield from self.blob_parser.lazy_parse(blob)
+
+    def load(self) -> List[Document]:
+        """Load all documents."""
+        return list(self.lazy_load())
+
+    def load_and_split(
+        self, text_splitter: Optional[TextSplitter] = None
+    ) -> List[Document]:
+        """Load all documents and split them into sentences."""
+        raise NotImplementedError(
```

Comment:
why can't we load and split (when text_splitter is not None)?

Reply:
we can tackle it in the initializer and have it available through any the lazy method of the load method, I don't think we need another method to do the same thing, we could add it for backwards compatibility potentially but I want to add something that can simply process text.